### PR TITLE
add legal repo name check

### LINKED ISSUE #10460: Helm repo_name naming conventions
Helm allows you to add a repo with the following naming conventions

`helm repo add hub/library https://hub/chartrepo/library`

It works fine with the search option

`helm search repo hub/library`

However when pulling down images, it fails
```
~> helm pull hub/library/elasticsearch
Error: chart "library/elasticsearch" matching not found in hub index (try 'helm repo update'): no chart found
```

This is because helm can't allow differentiate between repo_name and chart. However In that case it should avoid having the repo_name created with `/` in the first place itself.

### PR DESCRIPTION
**Pull request by yank1**
**What this PR does / why we need it**:

FIX The issue "Helm repo_name naming conventions" #10460, check repo name if contains "/"

closes #10460

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [*] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: cmd/helm/repo_add.go
Comment (zonggen) [inline cmd/helm/repo_add.go:180]:
```suggestion
	// Check if the repo name is legal
```

File: cmd/helm/repo_add_test.go
Comment (yxxhero) [inline cmd/helm/repo_add_test.go:152]: Unset the env var before func return.

File: cmd/helm/repo_add.go
Comment (yank1) [reply cmd/helm/repo_add.go]: Fixed typo

File: cmd/helm/repo_add_test.go
Comment (yank1) [reply cmd/helm/repo_add_test.go:154]: @yxxhero fixed

**Comment by yank1** [comment]
HI @zonggen & @yxxhero ， Thanks for code review . Is there more comments about the PR ?

**Comment by yxxhero** [comment]
@yank1  Thans for your PR. I think all is ok. This PR need core maintainer to approve.  @hickeyma

**Comment by yankay** [comment]
HI @hickeyma , would you please review it?

**Review by hickeyma — APPROVED** [review]
LGTM, thanks @yank1

**Comment by hickeyma** [comment]
Needs review from another @helm/helm-classic-maintainers to be merged.

**Comment by yankay** [comment]
HI, @scottrigby would you please review the PR ? Thanks very much.

**Review by mattfarina — APPROVED** [review]

**Comment by yankay** [comment]
@mattfarina, thank you

**Comment by moustafab** [comment]
I think this is technically a breaking change which has been released in a patch. It's subtle but it took some time for me to figure out why our ArgoCD diffs started failing all of the sudden.

For example:
```
time="2022-03-10T20:17:33Z" level=info msg="helm repo add https://strimzi.io/charts https://strimzi.io/charts" dir=charts/kafka-operator execID=1c29d
time="2022-03-10T20:17:33Z" level=error msg="`helm repo add https://strimzi.io/charts [https://strimzi.io/charts`](https://strimzi.io/charts%60) failed exit status 1: Error: repository name (https://strimzi.io/charts) contains '/', please specify a different name without '/'" execID=1c29d
time="2022-03-10T20:17:33Z" level=info msg=Trace args="[helm repo add https://strimzi.io/charts https://strimzi.io/charts]" dir=charts/kafka-operator operation_name="exec helm" time_ms=26.396266
```

**Comment by bacongobbler** [comment]
@moustafab can you please comment on the point raised in https://github.com/helm/helm/issues/10460#issue-1079420732? According to the bug report, repository names containing forward slashes do not work in all cases because it wasn't considered a valid repository name.

The example OP described was:

```
$ helm repo add hub/library https://hub/chartrepo/library
```

Works, but then you cannot pull down a chart:

```
$ helm pull hub/library/elasticsearch
Error: chart "library/elasticsearch" matching not found in hub index (try 'helm repo update'): no chart found
```

On the point of breaking changes, I think this falls under acceptable breaking changes. A repository name like `https://strimzi.io/charts` wasn't considered a valid repository name, and we never considered that with `helm repo add`'s original design. However, if you can demonstrate that this repository name works under all normal conditions, perhaps we should grandfather it in as a "hack".

**Pull request merged**
